### PR TITLE
Typeahead

### PR DESCRIPTION
**Pull request by hasznosl**
Checklist:

- [ ] Test added / Snapshots updated
- [ ] Documentation added / updated

### CODE REVIEW

File: src/organisms/Typeahead.jsx [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+import PropTypes from 'prop-types'
+import { View, ListSpinner, Text, ListItem, Icon } from '../'
+import { css } from 'glamor'
+import { ColorPalette } from '@allthings/colors'
+
+class Typeahead extends React.Component {
+  static propTypes = {
+    configuration: PropTypes.object,
+    width: PropTypes.string,
+    selectCallback: PropTypes.func,
```
Comment (mosch) [inline src/organisms/Typeahead.jsx]: this should be called "onSelect"

File: src/organisms/Typeahead.jsx [hunk 2]
```diff
@@ -0,0 +1,241 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { View, ListSpinner, Text, ListItem, Icon } from '../'
+import { css } from 'glamor'
+import { ColorPalette } from '@allthings/colors'
+
+class Typeahead extends React.Component {
+  static propTypes = {
+    configuration: PropTypes.object,
```
Comment (mosch) [inline src/organisms/Typeahead.jsx]: Try to use PropTypes.shape to describe complex configurations.

File: src/organisms/Typeahead.jsx [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    width: PropTypes.string,
```
Comment (mosch) [inline src/organisms/Typeahead.jsx]: instead of having a width prop, you could simply allow to pass css like the other components do with the rest props

File: src/organisms/Typeahead.jsx [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    selectCallback: PropTypes.func,
+    // deleteElement: PropTypes.func,
+    multiselect: PropTypes.bool,
+    nrResults: PropTypes.number,
```
Comment (mosch) [inline src/organisms/Typeahead.jsx]: There is a better name for this. Actually, I think this can be opinionated by the component, i don't really see a need to make it configurable.

File: src/organisms/Typeahead.jsx [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+  componentDidMount() {
+    document.addEventListener('mousedown', this.handleClickOutside)
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('mousedown', this.handleClickOutside)
+  }
+
+  refetch = input => {
+    this.setState({ loading: true })
```
Comment (mosch) [inline src/organisms/Typeahead.jsx]: Instead of this, you need to have a property that the implementation can use to let the component know the results. S.th. like "onRequestResults()" and fetching and returing the list will be up to the implementation.

File: src/organisms/Typeahead.jsx [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+          <View
+            style={{
+              position: 'absolute',
+              zIndex: '2200',
+              cursor: 'pointer',
+              top: 18,
+              right: 10,
+            }}
+          >
+            <Icon
```
Comment (mosch) [inline src/organisms/Typeahead.jsx]: Not sure if we need this actually. We don't use a clear Icon anywhere else, is there a good reason to start this here?

File: stories/CollapsibleStory.jsx [hunk 7]
```diff
@@ -0,0 +1,178 @@
+import React from 'react'
```
Comment (mosch) [inline stories/CollapsibleStory.jsx]: Did this file slip into the PR? This seems unrelated to me.

Comment (hasznosl) [reply stories/CollapsibleStory.jsx]: I did a small refactoring on the storybook index and this refactoring included another component's story

File: src/organisms/Typeahead.jsx (hunk 6 above)
Comment (fivenp) [reply src/organisms/Typeahead.jsx]: I actually like the clear icon idea... 
In this case it makes even more sense but it would also be a good addition for TextInput in the future...

Comment (fivenp) [reply src/organisms/Typeahead.jsx]: Just take care with the zIndex... It might be too high (even staying on top of modals although the component is NOT inside the modal)

File: src/organisms/Typeahead.jsx [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+  handleClickOutside = e => {
+    if (this.listRef.current && !this.listRef.current.contains(e.target)) {
+      this.setState({ showResults: false, focusIndex: -1 })
+    }
+  }
+
+  clearInput = () => this.setState({ input: '' })
+
+  onKeyDown = e => {
+    e.stopPropagation()
```
Comment (dan-lee) [inline src/organisms/Typeahead.jsx]: Consider [`Event.preventDefault`](https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault) on arrow keys instead of stopping the propagation.

File: src/organisms/Typeahead.jsx [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+            />
+          </View>
+        )}
+        <input
+          onChange={this.handleInputChange}
+          onFocus={this.onInputFocus}
+          value={input}
+          ref={this.inputRef}
+          tabIndex={tabIndex}
+          {...css({ height: 43, textIndent: 22, border: 'none' })}
```
Comment (fivenp) [inline src/organisms/Typeahead.jsx]: See above... If we don't use our own `<TextInput>` we should apply the same styles (fonts, textIndent:15, etc etc) to this `<input>`

File: src/organisms/Typeahead.jsx [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+                      this.handleSelectItem({
+                        value: option[configuration.value],
+                        label: option[configuration.label],
+                      })
+                    }
+                  }}
+                  {...css({
+                    width,
+                    backgroundColor: isSelected
+                      ? ColorPalette.lightGrey
```
Comment (fivenp) [inline src/organisms/Typeahead.jsx]: ` alpha(ColorPalette.background.light, 0.25)`

File: src/organisms/Typeahead.jsx [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+                      : ColorPalette.white,
+                    ':focus': {
+                      backgroundColor: isSelected
+                        ? ColorPalette.lightGreyIntense
```
Comment (fivenp) [inline src/organisms/Typeahead.jsx]: `alpha(ColorPalette.background.light, 0.40)`

File: src/organisms/Typeahead.jsx [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+                        value: option[configuration.value],
+                        label: option[configuration.label],
+                      })
+                    }
+                  }}
+                  {...css({
+                    width,
+                    backgroundColor: isSelected
+                      ? ColorPalette.lightGrey
+                      : ColorPalette.white,
```
Comment (fivenp) [inline src/organisms/Typeahead.jsx]:
```
':hover': {
  backgroundColor: alpha(ColorPalette.background.light, 0.25)
},
```
